File: runtimes/clr/src/rt/Space.cs
using System.Runtime.InteropServices;

namespace Flint.Rt;

/// Raw memory: one flat, byte-addressed space the whole runtime lives inside.
/// Ported from `runtime/src/mem.rs`.
///
/// ## Why unmanaged, and not a byte[] or long[]
///
/// A `byte[]` caps at 2 GB and forces byte-at-a-time access; a `long[]` reaches
/// ~17 GB but lives on the managed heap, counting against the GC and needing
/// one contiguous allocation.
///
/// `NativeMemory` has neither problem, and .NET is the better of the two hosts
/// here: `unsafe` gives real pointer arithmetic with no bounds check at all,
/// where the JVM's `MemorySegment` keeps one the JIT has to hoist. This is the
/// exact analogue of wasm's linear memory, which is the point -- it is the only
/// file in the port that differs from the Rust in more than syntax, because it
/// is the only one that touches the platform.
public sealed unsafe class Space : System.IDisposable {
    public const long Page = 65536;

    private byte* _base;

    /// How much has been handed out, and how much there is. `Addr`-wide,
    /// because a space can exceed 4 GB.
    public long InUse;
    public long Reserved;

    public Space(long bytes) {
        _base = (byte*) NativeMemory.AllocZeroed((nuint) bytes);
        if (_base == null) throw new System.OutOfMemoryException("flint: could not reserve the heap");
        Reserved = bytes;
        // Address 0 is never a valid object.
        InUse = Page;
    }

    public void Dispose() {
        if (_base != null) { NativeMemory.Free(_base); _base = null; }
    }

    public static long AlignUp(long n, long a) => (n + a - 1) & ~(a - 1);

    /// Hand out `len` bytes. 0 means full -- never an exception, because the
    /// caller turns it into a catchable flint error.
    public long Take(long len) {
        len = AlignUp(len, Page);
        if (InUse + len > Reserved) return 0;
        long a = InUse;
        InUse += len;
        return a;
    }

    public int ReadU32(long addr) => *(int*)(_base + addr);
    public void WriteU32(long addr, int v) => *(int*)(_base + addr) = v;
    public long ReadU64(long addr) => *(long*)(_base + addr);
    public void WriteU64(long addr, long v) => *(long*)(_base + addr) = v;

    // --- atomic word access -------------------------------------------------
    //
    // Mirrors the Rust `Space::atomic_load` / `cas`. There was an
    // `atomic_store` in all three and nothing ever called it: publication
    // goes through `cas`, which also runs the WRITE BARRIER, so a bare
    // atomic store to a slot would skip it. Removed 2026-09-23. Every
    // object is 8-aligned -- `SizeFor` rounds each layout to a multiple of 8 --
    // so `SlotAddr`, which is `base + 8 + i * 8`, is always aligned, which is
    // what makes an interlocked operation on it legal.
    //
    // These exist for ONE thing: a port's inbox, where two executors reserve
    // and publish without a lock (`DECISIONS.md#drivers`). Nothing else in the
    // heap is written by two threads at once -- a collection is stop-the-world
    // at a safepoint -- so ordinary slots need no synchronisation.
    public long AtomicLoad(long addr) =>
        System.Threading.Volatile.Read(ref *(long*)(_base + addr));

    /// Compare-and-swap. True when this thread won the slot.
    public bool Cas(long addr, long want, long next) =>
        System.Threading.Interlocked.CompareExchange(
            ref *(long*)(_base + addr), next, want) == want;
    public int ReadU8(long addr) => _base[addr];
    public void WriteU8(long addr, int v) => _base[addr] = (byte) v;

    public byte[] Bytes(long addr, int len) {
        var out_ = new byte[len];
        Marshal.Copy((System.IntPtr)(_base + addr), out_, 0, len);
        return out_;
    }

    public void WriteBytes(long addr, byte[] src) =>
        Marshal.Copy(src, 0, (System.IntPtr)(_base + addr), src.Length);

    public void CopyWithin(long from, long to, long len) =>
        System.Buffer.MemoryCopy(_base + from, _base + to, len, len);

    public void Zero(long addr, long len) =>
        NativeMemory.Clear(_base + addr, (nuint) len);
}

File: runtimes/jvm/src/com/flint/rt/Space.java
package com.flint.rt;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;

/// Raw memory: one flat, byte-addressed space the whole runtime lives inside.
/// Ported from `runtime/src/mem.rs`.
///
/// ## Why off-heap, and not a `byte[]` or a `long[]`
///
/// A `byte[]` caps at 2 GB (the index is an `int`) and forces byte-at-a-time
/// access. A `long[]` reaches ~17 GB but lives on the Java heap: it counts
/// against `-Xmx`, needs one contiguous allocation, and G1 treats it as a
/// humongous object it must then work around.
///
/// A `MemorySegment` from the FFM API (final in JDK 22) has none of that. It is
/// `long`-addressed, off-heap, invisible to the host collector, and its
/// accessors are JIT intrinsics. It is also the exact analogue of wasm's linear
/// memory, which is the point: this file is the ONLY one in the port that
/// differs from the Rust in more than syntax, because it is the only one that
/// touches the platform.
///
/// ## Lifetime is ours
///
/// An `Arena` owns the mapping, and closing it frees the heap immediately
/// rather than whenever the host collector gets round to it. That is the right
/// shape for a sandbox: dropping one should release its memory now.
public final class Space implements AutoCloseable {
    /// Unaligned layouts throughout. The Rust reads with `read_unaligned`, and
    /// an object header is 8-byte aligned while a string's bytes are not.
    private static final ValueLayout.OfLong I64 = ValueLayout.JAVA_LONG_UNALIGNED;
    private static final ValueLayout.OfInt I32 = ValueLayout.JAVA_INT_UNALIGNED;
    private static final ValueLayout.OfByte I8 = ValueLayout.JAVA_BYTE;

    public static final long PAGE = 65536;

    private final Arena arena;
    private final MemorySegment mem;

    /// How much of the space has been handed out, and how much there is.
    /// `Addr`-wide, because a space can exceed 4 GB now.
    public long inUse;
    public long reserved;

    public Space(long bytes) {
        this.arena = Arena.ofShared();
        this.mem = arena.allocate(bytes, 8);
        this.reserved = bytes;
        // Address 0 is never a valid object, so nothing is handed out from it.
        this.inUse = PAGE;
    }

    @Override public void close() { arena.close(); }

    public static long alignUp(long n, long a) { return (n + a - 1) & ~(a - 1); }

    /// Hand out `len` bytes. 0 means the space is full -- never an exception,
    /// because the caller turns it into a catchable flint error and a Java
    /// exception here would unwind past the frame that knows what failed.
    public long take(long len) {
        len = alignUp(len, PAGE);
        if (inUse + len > reserved) return 0;
        long a = inUse;
        inUse += len;
        return a;
    }

    public int readU32(long addr) { return mem.get(I32, addr); }
    public void writeU32(long addr, int v) { mem.set(I32, addr, v); }
    public long readU64(long addr) { return mem.get(I64, addr); }
    public void writeU64(long addr, long v) { mem.set(I64, addr, v); }

    // --- atomic word access -------------------------------------------------
    //
    // Mirrors the Rust `Space::atomic_load` / `atomic_store` / `cas`. Every
    // object is 8-aligned -- `sizeFor` rounds each layout to a multiple of 8 --
    // so `slotAddr`, which is `base + 8 + i * 8`, is always aligned and these
    // are legal. A `VarHandle` over a MemorySegment supports the atomic access
    // modes only for aligned addresses, which is the same requirement.
    //
    // These exist for ONE thing: a port's inbox, where two executors reserve
    // and publish without a lock (`doc/decisions/0028`). Nothing else in the
    // heap is written by two threads at once -- a collection is stop-the-world
    // at a safepoint -- so ordinary slots need no synchronisation.
    private static final java.lang.invoke.VarHandle I64_ATOMIC =
        ValueLayout.JAVA_LONG.varHandle();

    public long atomicLoad(long addr) {
        return (long) I64_ATOMIC.getVolatile(mem, addr);
    }

    public void atomicStore(long addr, long v) {
        I64_ATOMIC.setVolatile(mem, addr, v);
    }

    /// Compare-and-swap. True when this thread won the slot.
    public boolean cas(long addr, long want, long next) {
        return I64_ATOMIC.compareAndSet(mem, addr, want, next);
    }
    public int readU8(long addr) { return mem.get(I8, addr) & 0xFF; }
    public void writeU8(long addr, int v) { mem.set(I8, addr, (byte) v); }

    public byte[] bytes(long addr, int len) {
        byte[] out = new byte[len];
        MemorySegment.copy(mem, I8, addr, out, 0, len);
        return out;
    }

    public void writeBytes(long addr, byte[] src) {
        MemorySegment.copy(src, 0, mem, I8, addr, src.length);
    }

    public void copyWithin(long from, long to, long len) {
        MemorySegment.copy(mem, from, mem, to, len);
    }

    public void zero(long addr, long len) {
        mem.asSlice(addr, len).fill((byte) 0);
    }
}
